File: 0270_closest_bst_value/closest_value.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def closest_value(root: TreeNode, target: float) -> int:
    """
    Time: O(log(n))
    Space: O(1)
    """
    res = root.val
    while root:
        if abs(root.val - target) < abs(res - target):
            res = root.val
        root = root.left if target < root.val else root.right
    return res

File: 0270_closest_bst_value/closest_value.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Time: O(log(n))
// Space: O(1)
func closestValue(root *TreeNode, target float64) int {
	res := root.Val
	for root != nil {
		if math.Abs(target-float64(root.Val)) < math.Abs(target-float64(res)) {
			res = root.Val
		}
		if target > float64(root.Val) {
			root = root.Right
		} else {
			root = root.Left
		}
	}
	return res
}

func main() {

}
